url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/tools/RegressionValidator.java

```diff
@@ -0,0 +1,114 @@
+ * JANNLab Neural Network Framework for Java
+ * Copyright (C) 2012-2013 Sebastian Otte
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package de.jannlab.tools;
+
+import de.jannlab.Net;
+import de.jannlab.data.Sample;
+
+/**
+ * This class allows to simply evaluate regression tasks. Therefore
+ * a threshold is given. A sample for which the reference network
+ * produces a error < threshold is counted as recognized. After 
+ * testing a couple of samples the recognition ratio can be extracted.  
+ * <br></br>
+ * @author Sebastian Otte
+ */
+public class RegressionValidator {
+    
+    /**
+     * The reference network.
+     */
+    private Net net;
+    /**
+     * Number of tested samples.
+     */
+    private int ctr = 0;
+    /**
+     * Number of well recognized samples.
+     */
+    private int goodctr = 0;
+    /**
+     * The regression theshold.
+     */
+    private double threshold = 0.001;
+    
+    /**
+     * Returns the internal counter.
+     */
+    public void reset() {
+        this.ctr     = 0;
+        this.goodctr = 0;
+    }
+    
+    /**
+     * Creates an instance of RegressionValidator.
+     * @param net A given reference network.
+     * @param threshold The regression threshold.
+     */
+    public RegressionValidator(final Net net, final double threshold) {
+        this.net       = net;
+        this.threshold = threshold;
+    }
+    
+    /**
+     * Returns the current recognition rate. 100 = 1.0
+     * @return The current recognition rate.
+     */
+    public double ratio() {
+        return ((double)this.goodctr) / ((double)this.ctr);
+    }
+    
+    /**
+     * Applies a given sample to the reference network for only given selection.
+     * <br></br>
+     * @param sample An instance of Sample.
+     * @param features A selection as array of indices.
+     * @return The output error of the network after processing the sample.
+     */
+    public double apply(final Sample sample, final int[] features) {
+        this.net.reset();
+        this.ctr++;
+        //
+        // feed net with sample (discard error).
+        //
+        double error = 0.0;
+        if (features == null) {
+            error = NetTools.performForward(this.net, sample);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;sample&#39; (index: 0 | type: Sample). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;features&#39;. Other methods validating the parameter: &#39;apply&#39;. The same parameter type is validated here for example: url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/tools/ClassificationValidator.java:114. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).